File: leetcode-top-practice-problems/1520-maximum-number-of-non-overlapping-substrings/solution.go
package p1520

func maxNumOfSubstrings(s string) []string {
	first, last := [26]int{}, [26]int{}
	for i := range first {
		first[i] = len(s)
		last[i] = -1
	}
	for i := 0; i < len(s); i++ {
		c := int(s[i] - 'a')
		if i < first[c] {
			first[c] = i
		}
		last[c] = i
	}

	answer := make([]string, 0)
	previousEnd := -1
	for start := 0; start < len(s); start++ {
		c := int(s[start] - 'a')
		if first[c] != start {
			continue
		}
		end := last[c]
		valid := true
		for i := start; i <= end; i++ {
			x := int(s[i] - 'a')
			if first[x] < start {
				valid = false
				break
			}
			if last[x] > end {
				end = last[x]
			}
		}
		if !valid {
			continue
		}

		if start > previousEnd {
			answer = append(answer, s[start:end+1])
		} else {
			answer[len(answer)-1] = s[start : end+1]
		}
		previousEnd = end
	}
	return answer
}

File: leetcode-top-practice-problems/1520-maximum-number-of-non-overlapping-substrings/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<String> maxNumOfSubstrings(String s) {
        int[] first = new int[26];
        int[] last = new int[26];
        Arrays.fill(first, s.length());
        Arrays.fill(last, -1);

        for (int i = 0; i < s.length(); i++) {
            int letter = s.charAt(i) - 'a';
            first[letter] = Math.min(first[letter], i);
            last[letter] = i;
        }

        List<String> answer = new ArrayList<>();
        int previousEnd = -1;
        for (int start = 0; start < s.length(); start++) {
            int letter = s.charAt(start) - 'a';
            if (first[letter] != start) {
                continue;
            }

            int end = expandInterval(s, start, last, first);
            if (end == -1) {
                continue;
            }
            if (start > previousEnd) {
                answer.add(s.substring(start, end + 1));
            } else {
                answer.set(answer.size() - 1, s.substring(start, end + 1));
            }
            previousEnd = end;
        }
        return answer;
    }

    private int expandInterval(String s, int start, int[] last, int[] first) {
        int end = last[s.charAt(start) - 'a'];
        for (int i = start; i <= end; i++) {
            int letter = s.charAt(i) - 'a';
            if (first[letter] < start) {
                return -1;
            }
            end = Math.max(end, last[letter]);
        }
        return end;
    }
}
